### PR TITLE
#26: Added user guide

### LINKED ISSUE #26: Add user guide
# Background

We need a User Guide for this repo. It should contain  the following sections:
  * Setup
  * Usage 
  * Implementation of algorithms

### PR DESCRIPTION
**Pull request by umitbuyuksahin**
Fixes #26

### CODE REVIEW

File: doc/user_guide/user_guide.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+#### Install The Python Wheel Package
+
+The following command installs the package `advanced-analytics-framework` from [pypi](https://pypi.org):
+
+```bash
+pip install exasol-advanced-analytics-framework
+```
+
+### The Pre-built Script Language Container (SLC)
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:38]: I think, we need to try something here, if you look at _Implementation of Algorithms_ , you see a user can configure a UDF, maybe the class can be implemented in the UDF, then a prebuilt container would still make sense.

Comment (tkilias) [reply doc/user_guide/user_guide.md:38]: But, maybe it still makes sense, that this framework could be used as a dependency for a separate extension.

File: doc/user_guide/user_guide.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+- [Usage](#usage)
+- [Implementation of Algorithms](#implementation-of-algorithms)
+
+
+
+## Getting Started
+- Exasol DB
+  - The Exasol cluster must already be running with version 7.1 or later.
+  - DB connection information and credentials are needed.
+- TODO: Connection
```
Comment (tkilias) [inline doc/user_guide/user_guide.md]: Example for a bucketfs connection via create connection.

File: doc/user_guide/user_guide.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+- TODO: Algorithm implementation
```
Comment (tkilias) [inline doc/user_guide/user_guide.md]: obsolete

File: doc/user_guide/user_guide.md (hunk 2 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md]: See next push.

File: doc/user_guide/user_guide.md (hunk 1 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:38]: I think we first need to decide whether AAF should offer a pre-built SLC or not.

If yes, then we probably need to add this pre-built SLC or some logic in order to build it. 
This will probably be addressed in a separate ticket and hence out of scope for the current ticket. 

In any case, I propose to either remove or update the current section in the user guide until the decision and the potentially required changes are applied.

File: doc/user_guide/user_guide.md (hunk 3 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md]: Regarding the Algorithm implementation, I had the same impression.
For the connection: see next push.

File: doc/user_guide/user_guide.md (hunk 1 above)
Comment (tkilias) [reply doc/user_guide/user_guide.md:90]: if we offer an prebuilt SLC depends on, if my idea works or not. If it works, it makes sense, otherwise not. So, your task is probably trying it out. Built a custom UDF which calls the UDFQueryHandlerRunner with a custom UDFQueryHander.

Comment (ckunki) [reply doc/user_guide/user_guide.md:90]: Current experiments showed 
* UDF is in Python module `builtin`
* SLC could be uploaded to BucketFS of database in Docker Container

File: doc/user_guide/user_guide.md [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+```bash
+pip install exasol-advanced-analytics-framework
+```
+
+### Script Language Containers (SLCs)
+
+Exasol executes User Defined Functions (UDFs) in an isolated Script Language Container (SLCs).
+Running the AAF requires a custom SLC.
+
+#### Build the SLC
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:89]: I think, this section we can move to the developer guide, because we will release the prebuilt slc

File: doc/user_guide/user_guide.md [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    --db-pass <DB_PASSWORD> \
+    --bucketfs-name <BUCKETFS_NAME> \
+    --bucketfs-host <BUCKETFS_HOST> \
+    --bucketfs-port <BUCKETFS_PORT> \
+    --bucketfs-user <BUCKETFS_USER> \
+    --bucketfs-password <BUCKETFS_PASSWORD> \
+    --bucket <BUCKETFS_NAME> \
+    --path-in-bucket <PATH_IN_BUCKET> \
+    --language-alias "$LANGUAGE_ALIAS" \
+    --container-file "$SLC_FILE"
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:126]: the slc deployer should be able to download the prebuilt slc automatically from github, see https://github.com/exasol/transformers-extension/blob/f31b7ea91db5e560b6207b6efcba52eddfd4c7ce/exasol_transformers_extension/deployment/te_language_container_deployer.py#L11

--container-file is only fallback for special circumstances

File: doc/user_guide/user_guide.md [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    --db-pass <DB_PASSWORD> \
+    --schema <DB_SCHEMA> \
+    --language-alias "$LANGUAGE_ALIAS"
+```
+
+## Usage
+
+The entry point of this framework is `AAF_RUN_QUERY_HANDLER` script. This script is simply a query loop which is responsible for executing the implemented algorithm.
+
+This script takes the necessary parameters to execute the desired algorithm in string json format. The json input includes two main part:
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:148]:
```suggestion
This script takes the necessary parameters to execute the desired algorithm in string json format. The json input includes two main parts:
```

File: doc/user_guide/user_guide.md [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+
+* `query_handler` : Details of the algorithm implemented by user.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:150]:
```suggestion
* `query_handler` : Details of the algorithm implemented by the user.
```

File: doc/user_guide/user_guide.md [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    --dsn <DB_HOST:DB_PORT> \
+    --db-user <DB_USER> \
+    --db-pass <DB_PASSWORD> \
+    --schema <DB_SCHEMA> \
+    --language-alias "$LANGUAGE_ALIAS"
+```
+
+## Usage
+
+The entry point of this framework is `AAF_RUN_QUERY_HANDLER` script. This script is simply a query loop which is responsible for executing the implemented algorithm.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:146]: We might need to explain that an algorithm is implemented as QueryHandler and what a QueryHandler is.

File: doc/user_guide/user_guide.md [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+See [Implementation of Custom Algorithms](#implementation-of-custom-algorithms) for a complete example.
+
+### Parameters
+
+| Parameter                    | Required? | Description                                                                                                    |
+|------------------------------|-----------|----------------------------------------------------------------------------------------------------------------|
+| `<CLASS_NAME>`               | yes       | Name of the query handler class                                                                                |
+| `<CLASS_MODULE>`             | yes       | Module name of the query handler class <span style="color: red">(should we mention `builtins` here?)</span>    |
+| `<CLASS_PARAMETERS>`         | yes       | Parameters of the query handler class encoded as string                                                        |
+| `<UDF_NAME>`                 | -         | Name of Python UDF script including user-implemented algorithm                                                 |
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:187]:
```suggestion
| `<UDF_NAME>`                 | -         | Name of Python UDF script that contains the algorithm implemented by the user                                                |
```

File: doc/user_guide/user_guide.md [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+
+See [Implementation of Custom Algorithms](#implementation-of-custom-algorithms) for a complete example.
+
+### Parameters
+
+| Parameter                    | Required? | Description                                                                                                    |
+|------------------------------|-----------|----------------------------------------------------------------------------------------------------------------|
+| `<CLASS_NAME>`               | yes       | Name of the query handler class                                                                                |
+| `<CLASS_MODULE>`             | yes       | Module name of the query handler class <span style="color: red">(should we mention `builtins` here?)</span>    |
+| `<CLASS_PARAMETERS>`         | yes       | Parameters of the query handler class encoded as string                                                        |
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:186]: actually we should check, what happens when this is missing or a json object or array

File: doc/user_guide/user_guide.md [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+```
+
+See [Implementation of Custom Algorithms](#implementation-of-custom-algorithms) for a complete example.
+
+### Parameters
+
+| Parameter                    | Required? | Description                                                                                                    |
+|------------------------------|-----------|----------------------------------------------------------------------------------------------------------------|
+| `<CLASS_NAME>`               | yes       | Name of the query handler class                                                                                |
+| `<CLASS_MODULE>`             | yes       | Module name of the query handler class <span style="color: red">(should we mention `builtins` here?)</span>    |
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:185]: the builtins only comes into play when we implement the queryhandler in UDF. We should mention that in the example.

File: doc/user_guide/user_guide.md [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+| `<CLASS_PARAMETERS>`         | yes       | Parameters of the query handler class encoded as string                                                        |
+| `<UDF_NAME>`                 | -         | Name of Python UDF script including user-implemented algorithm                                                 |
+| `<UDF_DB_SCHEMA>`            | -         | Schema name where the UDF script is deployed                                                                   |
+| `<BUCKETFS_CONNECTION_NAME>` | yes       | BucketFS connection name to create temporary file outputs                                                      |
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:189]:
```suggestion
| `<BUCKETFS_CONNECTION_NAME>` | yes       | BucketFS connection name to create temporary bucketfs files                                                      |
```

File: doc/user_guide/user_guide.md [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+| `<BUCKETFS_DIRECTORY>`       | yes       | Directory in BucketFS for temporary file outputs                                                               |
+| `<TEMP_DB_SCHEMA>`           | yes       | Database Schema for temporary database objects, e.g. tables                                                    |
+
+# Implementation of Custom Algorithms
+
+Each algorithm should extend the `UDFQueryHandler` abstract class and then implement the following methods:
+* `start()`: This method is called at the first execution of the framework, that is, in the first iteration. It returns a result object: Either _Finish_ or _Continue_.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:196]:
```suggestion
* `start()`: This method is called at the first execution of the query hander, that is, in the first iteration. It returns a result object: Either _Finish_ or _Continue_.
```

File: doc/user_guide/user_guide.md [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+  * The _Finish_ result object contains the final result of the implemented algorithm.
+  * The _Continue_ object contains the query list that will be executed for the next state.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:198]:
```suggestion
  * The _Continue_ object contains the query list that will be executed before the next iteration and whose results are used as input fo the next iteraton.
```

File: doc/user_guide/user_guide.md [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+* `handle_query_result()`: This method is called at the following iterations to handle the return query.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:199]:
```suggestion
* `handle_query_result()`: This method is called at the following iterations to handle the result of the queries of the previous iteration.
```

File: doc/user_guide/user_guide.md [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:200]: We probably need to mention usage modes for the AAF,

1) Adhoc implementation of algorithms within a UDF
2) Building an extension
  - means python package which implements the query handler and an associated SLC which has the python package installed 
  - here we only give a link to PEC
  - we give an example how would need to AAF_RUN_QUERY_HANDLER in this case

File: doc/user_guide/user_guide.md [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+from exasol_advanced_analytics_framework.query_handler.context.query_handler_context import QueryHandlerContext
+from exasol_advanced_analytics_framework.query_result.query_result import QueryResult
+from exasol_advanced_analytics_framework.query_handler.result import Result, Continue, Finish
+from exasol_advanced_analytics_framework.query_handler.query.select_query import SelectQuery, SelectQueryWithColumnDefinition
+from exasol_data_science_utils_python.schema.column import Column
+from exasol_data_science_utils_python.schema.column_name import ColumnName
+from exasol_data_science_utils_python.schema.column_type import ColumnType
+
+
+-- proposal write parameter into temp table in schema ...
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:219]: maybe the following scenario

start
1. write parameter+"bucketfs"+time to temp bucketfs file
2. create temp table
3. insert parameter+"table-insert"+time to temp table
4. Return query select from temp table and concat "return"

handle result
1. read result
2. read temp bucketfs file
3. create table from result
4. insert temp bucketfs file content into new table
5. return new table name

File: doc/user_guide/user_guide.md [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+            query_list=query_list,
+            input_query=query_handler_return_query)
+
+    def handle_query_result(self, query_result: QueryResult) -> Union[Continue, Finish[str]]:
+        return_value = query_result.return_column
+        result = 2 ** return_value
+        return Finish(result=f"Assertion of the final result: 32 == {result}")
+
+import builtins
+builtins.CustomQueryHandler=CustomQueryHandler # required for pickle
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:245]: in a extra PR we need to add a function like register_query_handler_inside_udf and comment that this is only needed when you define the query handler insidde of a UDF

File: doc/user_guide/user_guide.md [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+}');
+```
+
+The figure below illustrates the execution of this algorithm implemented in class `ExampleQueryHandler`.
+* When method `start()` is called, it executes two queries and an additional `input_query` to obtain the next state.
+* After the first iteration is completed, the framework calls method `handle_query_result` with the `query_result` of the `input_query` of the previous iteration.
+
+In this example, the algorithm is finished at this state and returns 2<sup>_return value_</sup> as final result.
+
+![Sample Execution](../images/sample_execution.png "Sample Execution")
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:291]: With the extended scenario, we need to redo the image

File: doc/user_guide/user_guide.md (hunk 4 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:89]: See next push

File: doc/user_guide/user_guide.md (hunk 5 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:126]: Ok, I am sorry to maybe have created some confusion.
Yes: AAF deploy CLI has already been ported to the latest PEC version and is able to download the SLC file from the specified GitHub release.

File: doc/user_guide/user_guide.md (hunk 11 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:185]: See next push.

File: doc/user_guide/user_guide.md (hunk 10 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:186]: Using an array `"parameters": [ "bla-bla" ],` results in an unusable string containing a reference passed to the UDF: `table: 0x14823bd38580`.

Comment (ckunki) [reply doc/user_guide/user_guide.md:186]: I added some explanation to the user guide.

File: doc/user_guide/user_guide.md (hunk 16 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:200]: This sentence I don't understand. 

> we give an example how would need to AAF_RUN_QUERY_HANDLER in this case

File: doc/user_guide/udf-6.sql [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+        self._bfs_temp_file.write(self._sample_content("bucketfs"))
+
+    def _read_from_bucketfs(self) -> str:
+        buffer = BytesIO()
+        for chunk in self._bfs_temp_file.read():
+            buffer.write(chunk)
+        return buffer.getvalue().decode("utf-8")
+
+    def _write_to_db_table(self):
+        self.db_table_proxy().insert(self._sample_content("table-insert"))
```
Comment (tkilias) [inline doc/user_guide/udf-6.sql:49]: ahm, you need to construct an insert statement and return this as part of the Continue of the start method

File: doc/developer_guide/developer_guide.md [hunk 21]
```diff
@@ -1,8 +1,19 @@
...
+## Build the SLC
+
+If you want to build the AAF SLC then
+
+1. Download the [AAF Sources](https://github.com/exasol/advanced-analytics-framework/) from GitHub
+2. Install poetry
+3. Use the AAF developer commands for building and deploying custom SLCs into Exasol.
+
+```shell
+poetry run nox build_language_container
```
Comment (tkilias) [inline doc/developer_guide/developer_guide.md:16]:
```suggestion
poetry run nox -s build_language_container
```

File: doc/user_guide/user_guide.md (hunk 16 above)
Comment (tkilias) [reply doc/user_guide/user_guide.md:196]: > This sentence I don't understand.
> 
> > we give an example how would need to AAF_RUN_QUERY_HANDLER in this case

we give an example how would a user need to call AAF_RUN_QUERY_HANDLER in this case

**Comment by tkilias** [comment]
- After some discussions with @mkcorneli , I also recognized we need to give some examples for the algorithms you can implement with this framework (not actually implementation in the first place), hera are some
  - Scalable Apriori for frequent item set mining on large tables
  - Scalable Clustering, like k-means on large tables
  - Scalable Graph Algorithms that go beyond the capabilities of connect by, such as Iterative Map Reduce based PageRank

File: doc/user_guide/user_guide.md (hunk 8 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:127]: Proposal: Move to separate PR.

File: doc/user_guide/user_guide.md (hunk 17 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:215]: Proposal: Move to separate PR.

File: doc/user_guide/user_guide.md (hunk 19 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:287]: We will keep the Image for now, please fix typo "Contiue". 
Image will be updated in separate PR, anyway, accordingly to the extended scenario.

File: doc/user_guide/udf-6.sql (hunk 20 above)
Comment (ckunki) [reply doc/user_guide/udf-6.sql:49]: Will be addressed in separate PR on extended scenario.

File: doc/user_guide/user_guide.md (hunk 10 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md]: See also #196

File: doc/user_guide/user_guide.md (hunk 16 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:196]: Thanks - see next push.

File: doc/developer_guide/developer_guide.md [hunk 22]
```diff
@@ -1,8 +1,19 @@
...
+
+1. Download the [AAF Sources](https://github.com/exasol/advanced-analytics-framework/) from GitHub
+2. Install poetry
+3. Use the AAF developer commands for building and deploying custom SLCs into Exasol.
+
+```shell
+poetry run nox -s build_language_container
+```
+
+See [Install the SLC](../user_guide/user_guide.md#install-the-slc).
```
Comment (tkilias) [inline doc/developer_guide/developer_guide.md:19]: We probably also need to explain how to run tests

File: doc/images/system_design_diagram.drawio
Comment (tkilias) [inline doc/images/system_design_diagram.drawio:1]: Why did you delete this?

File: doc/user_guide/user_guide.md [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+
+The latest version of AAF can be obtained from [pypi](https://pypi.org), see also the [releases on GitHub](https://github.com/exasol/advanced-analytics-framework/releases):
+
+```bash
+pip install exasol-advanced-analytics-framework
+```
+
+### Script Language Container (SLC)
+
+Exasol executes User Defined Functions (UDFs) in an isolated Container called _SLC_.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:74]:
```suggestion
Exasol executes User Defined Functions (UDFs) in a isolated Container whose root filesystem is derived Script Language Container (SLC).
```

Maybe add a link to SLC repo

File: doc/user_guide/user_guide.md [hunk 24]
```diff
@@ -0,0 +1,281 @@
...
+    --db-pass <DB_PASSWORD> \
+    --bucketfs-name <BUCKETFS_NAME> \
+    --bucketfs-host <BUCKETFS_HOST> \
+    --bucketfs-port <BUCKETFS_PORT> \
+    --bucketfs-user <BUCKETFS_USER> \
+    --bucketfs-password <BUCKETFS_PASSWORD> \
+    --bucket <BUCKETFS_NAME> \
+    --path-in-bucket <PATH_IN_BUCKET> \
+    --version <VERSION> \
+    --language-alias "$LANGUAGE_ALIAS"
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:95]: Hm, using $ and <> looks weird, you probably want to show that the language alias is needed in two places

File: doc/user_guide/user_guide.md [hunk 25]
```diff
@@ -0,0 +1,281 @@
...
+
+### Deployment Options
+
+Using the AAF requires to implement a custom algorithm using one of the following alternatives
+* Adhoc implementation within a UDF
+* Build a custom extension
+
+#### Building a custom extension
+
+* Create a python package implementing the query handler.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:174]:
```suggestion
* Create a python package that depends on the python package of the AAF and that implements the query handler.of the custom algorithm.
```

File: doc/user_guide/user_guide.md [hunk 26]
```diff
@@ -0,0 +1,281 @@
...
+* Create an associated SLC which has the python package installed
+  * GitHub repository [python-extension-common](https://github.com/exasol/python-extension-common/) provides more detailed documentation and automation.
+* Leave out entry `udf` from the json input to use the default UDF. <!--
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:177]: Your version is better, you can remove the original comment

File: doc/user_guide/user_guide.md [hunk 27]
```diff
@@ -0,0 +1,281 @@
...
+-->
+* Values `<CLASS_MODULE>` and `<CLASS_NAME>` must reflect the _module_ and _class name_ of the `QueryHandler` implemented in the custom SLC.
+
+
+### Implementation of the Custom Algorithm
+
+Each algorithm should extend the `UDFQueryHandler` abstract class and then implement the following methods:
+* `start()`: This method is called at the first execution of the query hander, that is, in the first iteration. It returns a result object: Either _Finish_ or _Continue_.
+  * The _Finish_ result object contains the final result of the implemented algorithm.
+  * The _Continue_ object contains the query list that will be executed before the next iteration and whose results are used as input fo the next iteraton.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:189]:
```suggestion
  * The _Continue_ object contains the query list that will be executed before the next iteration and whose results are used as input for the next iteraton.
```

File: doc/user_guide/user_guide.md [hunk 28]
```diff
@@ -0,0 +1,281 @@
...
+            "directory": "temp"
+        },
+        "schema_name": "TEMP_SCHEMA"
+    }
+}');
+```
+
+The figure below illustrates the execution of this algorithm implemented in class `ExampleQueryHandler`.
+* When method `start()` is called, it executes two queries and an additional `input_query` to obtain the next state.
+* After the first iteration is completed, the framework calls method `handle_query_result` with the `query_result` of the `input_query` of the previous iteration.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:277]:
```suggestion
* After the first iteration is completed, the framework calls method the `handle_query_result` with the `query_result` of the `input_query` of the previous iteration.
```

File: doc/user_guide/user_guide.md [hunk 29]
```diff
@@ -0,0 +1,281 @@
...
+
+In this example, the algorithm is finished at this state and returns 2<sup>_return value_</sup> as final result.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:279]: It actually returns

Finish(result=f"Assertion of the final result: 32 == {result}

Comment (tkilias) [reply doc/user_guide/user_guide.md:279]: Code and image are inconsistent, please adapt the code to the image for the moment

File: doc/developer_guide/developer_guide.md (hunk 22 above)
Comment (ckunki) [reply doc/developer_guide/developer_guide.md:19]: See next push

File: doc/user_guide/user_guide.md (hunk 24 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:95]: Yes, I already added a back-reference in the next example commandline, But see next push for
* consistently only use `$` references
* add a short rationale why environment variable `LANGUAGE_ALIAS` is set before

File: doc/user_guide/user_guide.md (hunk 26 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:177]: Thanks :pray:

File: doc/user_guide/user_guide.md (hunk 29 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:279]: See next push

File: doc/images/system_design_diagram.drawio
Comment (ckunki) [reply doc/images/system_design_diagram.drawio:1]: Because the drawio sources are embedded into the PNGs, hence separate files are redundant.
I deleted the files to make the repo smaller, and avoid inconsistencies.

File: doc/developer_guide/developer_guide.md [hunk 30]
```diff
@@ -1,8 +1,66 @@
...
+
+Installing the SLC ins described in the [AAF User Guide](../user_guide/user_guide.md#script-language-container-slc).
+
+## Running Tests
+
+AAF comes with different automated tests implemented in different programming languages and requiring different environments:
+
+| Language | Category                        | Database | Environment |
+|----------|---------------------------------|----------|-------------|
+| Python   | Unit tests                      | no       | poetry      |
```
Comment (tkilias) [inline doc/developer_guide/developer_guide.md:23]: are you sure, this might also test the lua script generation and this needs amalg.

File: doc/user_guide/user_guide.md [hunk 31]
```diff
@@ -0,0 +1,280 @@
...
+### Exasol database
+
+* The Exasol cluster must already be running with version 7.1 or later.
+* Database connection information and credentials are needed for the database itself as also for the BucketFS.
+
+### BucketFS Connection
+
+AAF employs some Lua scripts and User Defined Functions (UDFs).  The Lua scripts are orchestrating the UDFs while the UDFs are performing the actual analytic functions.
+
+AAF keeps a common state of execution and passes input data and results between Lua and UDFs via files in the Bucket File System (BucketFS) of the Exasol database.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:22]: This is basically a minimalistic explanation, maybe it should go into its own section

File: doc/user_guide/user_guide.md [hunk 32]
```diff
@@ -0,0 +1,280 @@
...
+
+The latest version of AAF can be obtained from [pypi](https://pypi.org), see also the [releases on GitHub](https://github.com/exasol/advanced-analytics-framework/releases):
+
+```bash
+pip install exasol-advanced-analytics-framework
+```
+
+### Script Language Container (SLC)
+
+Exasol executes User Defined Functions (UDFs) in a isolated Container whose root filesystem is derived Script Language Container (SLC).
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:74]:
```suggestion
Exasol executes User Defined Functions (UDFs) in a isolated Container whose root filesystem is derived from a Script Language Container (SLC).
```

File: doc/user_guide/user_guide.md [hunk 33]
```diff
@@ -0,0 +1,280 @@
...
+
+Running the AAF requires a custom SLC. The following command
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:76]:
```suggestion
Running the AAF requires a SLC. The following command
```

custom is here not correct, because it is not always customized, it could be also prebuilt.

File: doc/user_guide/user_guide.md [hunk 34]
```diff
@@ -0,0 +1,280 @@
...
+| `<CLASS_MODULE>`             | yes       | Module name of the query handler class                                        |
+| `<CLASS_PARAMETERS>`         | yes       | Parameters of the query handler class encoded as string                       |
+| `<UDF_NAME>`                 | -         | Name of Python UDF script that contains the algorithm implemented by the user |
+| `<UDF_DB_SCHEMA>`            | -         | Schema name where the UDF script is deployed                                  |
+| `<BUCKETFS_CONNECTION_NAME>` | yes       | BucketFS connection name which is used to create temporary bucketfs files     |
+| `<BUCKETFS_DIRECTORY>`       | yes       | Directory in BucketFS for the temporary bucketfs files                        |
+| `<TEMP_DB_SCHEMA>`           | yes       | Database Schema for temporary database objects, e.g. tables                   |
+
+Please take care to provide a string value for `<CLASS_PARAMETERS>`.  Simple data types like `float`, `int`, `bool` will be converted to a String, while a Json object or an array is represented as a string value with an unusable reference, e.g.  `table: 0x14823bd38580`.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:164]: We should that the json object problem will be fixed

File: doc/user_guide/user_guide.md [hunk 35]
```diff
@@ -0,0 +1,280 @@
...
+
+### Deployment Options
+
+Using the AAF requires to implement a custom algorithm using one of the following alternatives
+* Adhoc implementation within a UDF
+* Build a custom extension
+
+#### Building a custom extension
+
+* Create a python package that depends on the python package of the AAF and that implements the query handler.of the custom algorithm and its factory class.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:176]:
```suggestion
* Create a python package that depends on the python package of the AAF and that implements the query handler of the custom algorithm and its factory class.
```

File: doc/user_guide/user_guide.md [hunk 36]
```diff
@@ -0,0 +1,280 @@
...
+            "connection_name": "BFS_CON",
+            "directory": "temp"
+        },
+        "schema_name": "TEMP_SCHEMA"
+    }
+}');
+```
+
+The figure below illustrates the execution of this algorithm implemented in class `ExampleQueryHandler`.
+* When method `start()` is called, it executes two queries and an additional `input_query` to obtain the next state.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:275]:
```suggestion
* When method `start()` is called, it executes two queries and an additional `input_query` to obtain the input for the next .iteration.
```

File: doc/user_guide/user_guide.md [hunk 37]
```diff
@@ -0,0 +1,280 @@
...
+* After the first iteration is completed, the framework calls method the `handle_query_result` with the `query_result` of the `input_query` of the previous iteration.
+
+In this example, the algorithm is finished at this state and returns 2<sup>_return value_</sup> as final result.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:278]:
```suggestion
In this example, the algorithm is finished at this iteration and returns 2<sup>_return value_</sup> as final result.
```

File: doc/user_guide/user_guide.md (hunk 34 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:164]: See next push.

File: doc/user_guide/user_guide.md (hunk 31 above)
Comment (ckunki) [reply doc/user_guide/user_guide.md:22]: See next push.

File: doc/user_guide/user_guide.md [hunk 38]
```diff
@@ -0,0 +1,280 @@
...
+#### Building a custom extension
+
+* Create a python package that depends on the python package of the AAF and that implements the query handler of the custom algorithm and its factory class.
+* Create an associated SLC which has the python package installed
+  * GitHub repository [python-extension-common](https://github.com/exasol/python-extension-common/) provides more detailed documentation and automation.
+* Leave out entry `udf` from the json input to use the default UDF.
+* Values `<CLASS_MODULE>` and `<CLASS_NAME>` must reflect the _module_ and _class name_ of the `QueryHandler` implemented in the custom SLC.
+
+
+### Implementing a Custom Algorithm as Example Query Handler
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:183]:
```suggestion
### Example Implementation of a Custom Algorithm as Query Handler
```

File: doc/user_guide/user_guide.md [hunk 39]
```diff
@@ -0,0 +1,280 @@
...
+
+Each algorithm should extend the `UDFQueryHandler` abstract class and then implement the following methods:
+* `start()`: This method is called at the first execution of the query hander, that is, in the first iteration. It returns a result object: Either _Finish_ or _Continue_.
+  * The _Finish_ result object contains the final result of the implemented algorithm.
+  * The _Continue_ object contains the query list that will be executed before the next iteration and whose results are used as input for the next iteration.
+* `handle_query_result()`: This method is called at the following iterations to handle the result of the queries of the previous iteration.
+
+Here is an example class definition using an adhoc implementation within the UDF. The example uses the module `builtins` and dynamically adds ExampleQueryHandler and ExampleQueryHandlerFactory to it.
```
Comment (tkilias) [inline doc/user_guide/user_guide.md:191]: Add section Example

**Review by tkilias — APPROVED** [review]
